keydata.go

```diff
@@ -19,10 +19,488 @@
+	r := bytes.NewReader(c)
+
+	var sz uint16
+	if err := binary.Read(r, binary.BigEndian, &sz); err != nil {
+		return nil, nil, err
+	}
+
+	if sz > 0 {
+		key = make(DiskUnlockKey, sz)
+		if _, err := r.Read(key); err != nil {
+			return nil, nil, err
+		}
+	}
+
+	if err := binary.Read(r, binary.BigEndian, &sz); err != nil {
+		return nil, nil, err
+	}
+
+	if sz > 0 {
+		auxKey = make(AuxiliaryKey, sz)
+		if _, err := r.Read(auxKey); err != nil {
+			return nil, nil, err
+		}
+	}
+
+	if r.Len() > 0 {
+		return nil, nil, fmt.Errorf("%v excess byte(s)", r.Len())
+	}
+
+	return
+}
+
 // AuthMode corresponds to a set of authentication mechanisms.
 type AuthMode uint32
 
 const (
 	AuthModeNone       AuthMode = 0
 	AuthModePassphrase AuthMode = 1 << iota
 )
+
+// KeyCreationData is the data required to create a new KeyData object.
+// It should be produced by a platform implementation.
+type KeyCreationData struct {
+	PlatformKeyData
+	PlatformName string // Name of the platform that produced this data
+
+	// AuxiliaryKey is a key used to authorize changes to the key data.
+	// It must match the key protected inside PlatformKeyData.EncryptedPayload.
+	AuxiliaryKey AuxiliaryKey
+
+	// SnapModelAuthHash is the digest algorithm used for HMACs of Snap
+	// device models, and also the digest algorithm used to produce the
+	// key digest.
+	SnapModelAuthHash crypto.Hash
+}
+
+// KeyID is the unique ID for a KeyData object. It is used to facilitate the
+// sharing of state between the early boot environment and OS runtime.
+type KeyID struct {
+	Name     string // Unique name for the KeyData
+	Revision uint64 // Revision of the KeyData
+}
+
+func (id KeyID) String() string {
+	return fmt.Sprintf("%s@%d", id.Name, id.Revision)
+}
+
+// KeyDataWriter is an interface used by KeyData to write the data to
+// persistent storage in an atomic way.
+type KeyDataWriter interface {
+	io.Writer
+	Commit() error
+}
+
+// KeyDataReader is an interface used to read and decode a KeyData
+// from persistent storage.
+type KeyDataReader interface {
+	io.Reader
+	ID() KeyID
+}
+
+type hashAlg struct {
+	crypto.Hash
+}
+
+func (a hashAlg) MarshalJSON() ([]byte, error) {
+	var s string
+
+	switch a.Hash {
+	case crypto.SHA1:
+		s = "sha1"
+	case crypto.SHA224:
+		s = "sha224"
+	case crypto.SHA256:
+		s = "sha256"
+	case crypto.SHA384:
+		s = "sha384"
+	case crypto.SHA512:
+		s = "sha512"
+	default:
+		return nil, fmt.Errorf("unknown has algorithm: %v", a.Hash)
+	}
+
+	return json.Marshal(s)
+}
+
+func (a *hashAlg) UnmarshalJSON(b []byte) error {
+	var s string
+	if err := json.Unmarshal(b, &s); err != nil {
+		return err
+	}
+
+	switch s {
+	case "sha1":
+		a.Hash = crypto.SHA1
+	case "sha224":
+		a.Hash = crypto.SHA224
+	case "sha256":
+		a.Hash = crypto.SHA256
+	case "sha384":
+		a.Hash = crypto.SHA384
+	case "sha512":
+		a.Hash = crypto.SHA512
+	default:
+		a.Hash = crypto.Hash(0)
+	}
+
+	return nil
+}
+
+type snapModelHMAC []byte
+
+type snapModelHMACList []snapModelHMAC
```

Comment:
This will probably end up a list of list of intergers or something but we really want base64 here I think.